aten/src/ATen/core/Generator.h

```diff
@@ -1,24 +1,88 @@
 #pragma once
 
-#include <ATen/core/ATenGeneral.h>
 #include <stdint.h>
+#include <memory>
+#include <mutex>
+#include <deque>
+#include <atomic>
+#include <typeinfo>
+#include <utility>
+
+#include <c10/util/Exception.h>
+#include <c10/util/C++17.h>
+#include <ATen/Device.h>
+
+/*
+* Generator note.
+* A Pseudo Random Number Generator (PRNG) is an engine that uses an algorithm to 
+* generate a seemingly random sequence of numbers, that may be later be used in creating 
+* a random distribution. Such an engine almost always maintains a state and requires a
+* seed to start off the creation of random numbers. Often times, users have
+* encountered that it could be beneficial to be able to create, retain, and destroy 
+* PRNG states and also be able to have control over the seed value.
+*
+* A Generator in ATen gives users the ability to read, write and modify a PRNG engine.
+* For instance, it does so by letting users seed a PRNG engine, fork the state of the
+* engine, etc.
+*
+* By default, there is one generator per device, and a device's generator is 
+* lazily created. A user can use the torch.Generator() api to create their own generator.
+* Currently torch.Generator() can only create a CPUGenerator.
+*/
 
 namespace at {
 
+/*
+* CloneableGenerator class based on CRTP pattern. It is a
+* helper class used for cloning Generator subclasses while 
+* preserving covariance in clone() method.
+*/
+template <typename Derived, typename Base>
```

Comment:
Is there ever a case when Base won't be Generator?

Reply:
Probably not. I could remove the Base template param, unless we just put this helper in a `Cloneable.h` in aten and reuse the cloning method :)

Reply:
Well, the little extra bit of generality doesn't seem to do any harm. Not convinced we should start putting this idiom everywhere though lol